File: server/excalibur_server/api/routes/merkle/migrate.py
from typing import Annotated
from uuid import UUID

from fastapi import Body, Depends, HTTPException, status
from pydantic import Base64Bytes, BaseModel
from sqlalchemy.orm import Session

from excalibur_server.api.routes.merkle import encrypted_router
from excalibur_server.src.auth.credentials import Credentials, get_credentials
from excalibur_server.src.crypto.merkle.enums import MerkleStatus
from excalibur_server.src.crypto.merkle.structures import AttestationBase
from excalibur_server.src.db.operations import (
    count_items_with_root,
    get_session,
    get_unverified,
    get_user_from_id,
)
from excalibur_server.src.db.tables import Attestation, FSItem, VaultState


class MigrationEntry(BaseModel):
    """
    The Merkle data for a single item, submitted during a migration.
    """

    node_hash: Base64Bytes
    "Keyed MAC of the subtree rooted at this item"
    content_mac: Base64Bytes | None = None
    "Keyed MAC binding the file's AEAD tags to its identity, or None for folders"


def _refresh_counts(session: Session, root_id: UUID, vault_state: VaultState):
    """
    Recomputes a vault state's migration progress counters from the database.

    Must be given the caller's own session, since the counters have to take the caller's uncommitted
    writes into account.
    """

    vault_state.total_count = count_items_with_root(root_id, session)
    vault_state.migrated_count = vault_state.total_count - len(get_unverified(root_id, session))


@encrypted_router.post(
    "/migrate",
    name="Begin Migration",
    responses={
        status.HTTP_200_OK: {"description": "Migration started"},
        status.HTTP_409_CONFLICT: {"description": "Vault already has a Merkle tree"},
    },
    response_model=VaultState,
)
def begin_migration_endpoint(credentials: Annotated[Credentials, Depends(get_credentials)]):
    """
    Starts a migration of the authenticated user's vault to a Merkle tree.

    Moves the vault from the `NONE` status to the `MIGRATING` status, stamping the number of items
    that need hashes along the way.

    Note that ordinary, non-Merkle tree verified writes may continue during a migration. Any item
    written while the migration is in progress is treated as another item that needs to be filled
    in.
    """

    root_id = get_user_from_id(credentials.user_id).fsitem_id

    with get_session() as session, session.begin():
        vault_state = session.get(VaultState, root_id)
        if vault_state is None:
            vault_state = VaultState(root_id=root_id)
        elif vault_state.merkle_status != MerkleStatus.NONE:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=f"Vault is already in the '{vault_state.merkle_status.value}' state",
            )

        vault_state.merkle_status = MerkleStatus.MIGRATING
        vault_state.current_generation = 0
        _refresh_counts(session, root_id, vault_state)
        session.add(vault_state)

        result = vault_state.model_copy()

    return result


@encrypted_router.post(
    "/migrate/fill",
    name="Fill Migration Data",
    responses={
        status.HTTP_200_OK: {"description": "Chunk accepted"},
        status.HTTP_409_CONFLICT: {"description": "Vault is not migrating, or the chunk is invalid"},
    },
    response_model=VaultState,
)
def fill_migration_endpoint(
    credentials: Annotated[Credentials, Depends(get_credentials)],
    entries: Annotated[dict[UUID, MigrationEntry], Body(description="Mapping of item IDs to their Merkle data")],
):
    """
    Submits a chunk of Merkle data for an in-progress migration.

    This endpoint is resumable. That is,
    - chunks may be submitted in any order; and
    - re-submitting a chunk simply overwrites the data that was previously stored for those items.

    Identify data that needs migration using the `/merkle/dirty` endpoint.
    """

    root_id = get_user_from_id(credentials.user_id).fsitem_id

    with get_session() as session, session.begin():
        vault_state = session.get(VaultState, root_id)
        if vault_state is None or vault_state.merkle_status != MerkleStatus.MIGRATING:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Vault is not migrating")

        for item_id, entry in entries.items():
            item = session.get(FSItem, item_id)
            if item is None or item.root_id != root_id:
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT, detail=f"Item '{item_id}' is not in this vault"
                )

            if item.is_folder and entry.content_mac is not None:
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT, detail=f"Item '{item_id}' is a folder and has no content MAC"
                )
            if not item.is_folder and entry.content_mac is None:
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT, detail=f"Item '{item_id}' is a file and needs a content MAC"
                )

            item.node_hash = entry.node_hash
            item.content_mac = entry.content_mac
            item.version += 1
            session.add(item)

        _refresh_counts(session, root_id, vault_state)
        session.add(vault_state)

        result = vault_state.model_copy()

    return result


@encrypted_router.post(
    "/migrate/complete",
    name="Complete Migration",
    responses={
        status.HTTP_200_OK: {"description": "Migration completed"},
        status.HTTP_409_CONFLICT: {"description": "Vault is not migrating, or the attestation is invalid"},
    },
    response_model=Attestation,
)
def complete_migration_endpoint(
    credentials: Annotated[Credentials, Depends(get_credentials)],
    attestation: Annotated[AttestationBase, Body(description="The generation 1 attestation for the vault")],
):
    """
    Completes an in-progress migration, moving the vault to the `ACTIVE` status.

    Every item in the vault must have its Merkle data filled in before this is run.
    """

    root_id = get_user_from_id(credentials.user_id).fsitem_id

    # Structural checks on the submitted attestation
    if attestation.generation != 1:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="The migration attestation must be for generation 1"
        )
    if attestation.prev_root_hash is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT, detail="The migration attestation must not chain to a previous root"
        )

    new_attestation = Attestation.from_base(attestation, root_id)
    result = new_attestation.model_copy()

    with get_session() as session, session.begin():
        vault_state = session.get(VaultState, root_id)
        if vault_state is None or vault_state.merkle_status != MerkleStatus.MIGRATING:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail="Vault is not migrating")

        unverified = get_unverified(root_id, session)
        if unverified:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT, detail=f"Migration is incomplete; {len(unverified)} items remain"
            )

        root = session.get(FSItem, root_id)
        if root is None or root.node_hash != attestation.root_hash:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Attestation's root hash does not match the vault's root node hash",
            )

        session.add(new_attestation)

        vault_state.merkle_status = MerkleStatus.ACTIVE
        vault_state.current_generation = new_attestation.generation
        _refresh_counts(session, root_id, vault_state)
        session.add(vault_state)

    return result

File: app/packages/main/src/api/merkle/migrate.ts
import ExEF from "@lib/crypto/exef";
import { Attestation, AttestationBase, MigrationEntry, VaultState } from "@lib/merkle/structures";

import { popFetch } from "@api/fetch";

import { AuthProvider } from "@components/auth/context";

import { AttestationBaseWire, AttestationWire } from "./attestation";
import { VaultStateWire } from "./state";

/**
 * Begins migrating the current user's vault to a Merkle tree.
 *
 * @param auth the current authentication provider
 * @returns a promise which resolves to an object with a success boolean and optionally an error
 *      message or the vault state
 */
export async function beginMigration(
    auth: AuthProvider,
): Promise<{ success: boolean; error?: string; state?: VaultState }> {
    const response = await popFetch(`${auth.serverInfo!.apiURL}/merkle/migrate`, auth.authInfo!.key!, {
        method: "POST",
        headers: { Authorization: `Bearer ${auth.getToken()}` },
    });
    switch (response.status) {
        case 200:
            break; // Continue with normal flow
        case 401:
            return { success: false, error: "Unauthorized" };
        case 409:
            return { success: false, error: "Vault already has a Merkle tree" };
        default:
            return { success: false, error: "Unknown error" };
    }

    const state = (await new ExEF(auth.authInfo!.key!).decryptResponse<VaultStateWire>(response))!;
    return {
        success: true,
        state: {
            rootID: state.root_id,
            merkleStatus: state.merkle_status,
            currentGeneration: state.current_generation,
            migratedCount: state.migrated_count,
            totalCount: state.total_count,
        },
    };
}

interface MigrationEntryWire {
    node_hash: string;
    content_mac: string | null;
}

/**
 * Submits a chunk of Merkle data for an in-progress migration.
 *
 * This is resumable:
 * - chunks may be submitted in any order; and
 * - re-submitting a chunk simply overwrites the data that was previously stored for those items.
 *
 * @param auth the current authentication provider
 * @param entries mapping of item IDs to their Merkle data
 * @returns a promise which resolves to an object with a success boolean and optionally an error
 *      message or the updated vault state
 */
export async function fillMigration(
    auth: AuthProvider,
    entries: Record<string, MigrationEntry>,
): Promise<{ success: boolean; error?: string; state?: VaultState }> {
    const entriesWire: Record<string, MigrationEntryWire> = Object.fromEntries(
        Object.entries(entries).map(([id, entry]) => [
            id,
            {
                node_hash: entry.nodeHash,
                content_mac: entry.contentMAC,
            },
        ]),
    );

    const response = await popFetch(`${auth.serverInfo!.apiURL}/merkle/migrate/fill`, auth.authInfo!.key!, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${auth.getToken()}`,
            "Content-Type": "application/octet-stream",
            "X-Encrypted": "true",
            "X-Content-Type": "application/json",
        },
        // @ts-expect-error This is actually a valid body; its just that TS complains about it >:(
        body: await new ExEF(auth.authInfo!.key!, { version: 4 }).encrypt(
            Buffer.from(JSON.stringify(entriesWire), "utf-8"),
        ),
    });
    switch (response.status) {
        case 200:
            break; // Continue with normal flow
        case 401:
            return { success: false, error: "Unauthorized" };
        case 409:
            return { success: false, error: "Vault is not migrating, or the chunk is invalid" };
        default:
            return { success: false, error: "Unknown error" };
    }

    const state = (await new ExEF(auth.authInfo!.key!).decryptResponse<VaultStateWire>(response))!;
    return {
        success: true,
        state: {
            rootID: state.root_id,
            merkleStatus: state.merkle_status,
            currentGeneration: state.current_generation,
            migratedCount: state.migrated_count,
            totalCount: state.total_count,
        },
    };
}

/**
 * Completes an in-progress migration, moving the vault to the `active` status.
 *
 * Every item in the vault must have its Merkle data filled in (via {@link fillMigration}) before
 * this is called.
 *
 * @param auth the current authentication provider
 * @param attestation the generation-1 attestation for the vault
 * @returns a promise which resolves to an object with a success boolean and optionally an error
 *      message or the stored attestation
 */
export async function completeMigration(
    auth: AuthProvider,
    attestation: AttestationBase,
): Promise<{ success: boolean; error?: string; conflict?: boolean; attestation?: Attestation }> {
    const attestationWire: AttestationBaseWire = {
        generation: attestation.generation,
        root_hash: attestation.rootHash,
        prev_root_hash: attestation.prevRootHash,
        timestamp: attestation.timestamp,
        tag: attestation.tag,
    };

    const response = await popFetch(`${auth.serverInfo!.apiURL}/merkle/migrate/complete`, auth.authInfo!.key!, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${auth.getToken()}`,
            "Content-Type": "application/octet-stream",
            "X-Encrypted": "true",
            "X-Content-Type": "application/json",
        },
        // @ts-expect-error This is actually a valid body; its just that TS complains about it >:(
        body: await new ExEF(auth.authInfo!.key!, { version: 4 }).encrypt(
            Buffer.from(JSON.stringify(attestationWire), "utf-8"),
        ),
    });
    switch (response.status) {
        case 200:
            break; // Continue with normal flow
        case 401:
            return { success: false, error: "Unauthorized" };
        case 409:
            return { success: false, error: "Vault is not migrating, or the attestation is invalid", conflict: true };
        default:
            return { success: false, error: "Unknown error" };
    }

    const result = (await new ExEF(auth.authInfo!.key!).decryptResponse<AttestationWire>(response))!;
    return {
        success: true,
        attestation: {
            rootID: result.root_id,
            generation: result.generation,
            rootHash: result.root_hash,
            prevRootHash: result.prev_root_hash,
            timestamp: result.timestamp,
            tag: result.tag,
        },
    };
}
